backend/peephole/peephole_utils.ml

```diff
@@ -32,19 +27,39 @@ let get_cells cell size =
   assert (size > 0);
   get_cells' (DLL.next cell) (size - 1) [cell]
 
-let is_bitwise_op (op : Simple_operation.integer_operation) =
-  match op with Iand | Ior | Ixor | Ilsl | Ilsr | Iasr -> true | _ -> false
-  [@@ocaml.warning "-4"]
-
-let bitwise_shift_assert (imm1 : int) (imm2 : int) =
-  if imm1 < 0 || imm1 > Sys.int_size || imm2 < 0 || imm2 > Sys.int_size
-  then assert false
-  [@@inline]
-
-(* CR-someday gtulba-lecu: This is architecture specific and should be moved in
-   a different part of the compiler that is specific to the amd64 architecture.
-   This is fine for now as we only support amd64. *)
-let amd64_imm32_within_bounds imm1 imm2 op =
-  let imm = op imm1 imm2 in
-  Int32.to_int Int32.min_int <= imm && imm <= Int32.to_int Int32.max_int
-  [@@inline]
+let assert_within_range integer_operation imm =
+  if not (Arch.is_immediate_for_intop integer_operation imm)
+  then
+    Misc.fatal_errorf "Peephole: unexpected immediate %d for operation %s" imm
+      (Simple_operation.string_of_integer_operation integer_operation)
+
+let op_immediates integer_operation imm1 imm2 no_overflow op =
+  assert_within_range integer_operation imm1;
+  assert_within_range integer_operation imm1;
+  let res = op imm1 imm2 in
+  if no_overflow imm1 imm2 && Arch.is_immediate_for_intop integer_operation res
+  then Some (integer_operation, res)
+  else None
+  [@@inilne]
```

Comment:
```suggestion
  [@@inline]
```
(also, not 100% sure about the position;
I think we usually write `let[@inline] ...`.)